File: ios/loveaceios/Service/JWCService.swift
import Foundation
import SwiftSoup
import os

private let logger = Logger(subsystem: "tech.loveace.loveaceios", category: "JWCService")

actor JWCService {
    private let connection: AUFEConnection
    private var cachedTerms: UniResponse<[TermItem]>?
    private var termsCacheTimestamp: Date = .distantPast
    static let baseURL = "http://jwcxk2-aufe-edu-cn.vpn2.aufe.edu.cn:8118"
    private static let cacheTTL: TimeInterval = 30

    init(connection: AUFEConnection) {
        self.connection = connection
    }

    // MARK: - Academic Info

    func getAcademicInfo() async -> UniResponse<AcademicInfo> {
        do {
            let url = "\(Self.baseURL)/main/academicInfo?sf_request_type=ajax"
            let client = await connection.client!
            let (data, _) = try await client.post(url, formData: ["flag": ""], headers: [
                "Accept": "application/json, text/javascript, */*; q=0.01",
                "Referer": "\(Self.baseURL)/index.jsp",
                "X-Requested-With": "XMLHttpRequest"
            ])
            let body = String(data: data, encoding: .utf8) ?? ""
            guard !body.isEmpty else { throw ServiceError.emptyResponse }
            guard let jsonData = body.data(using: .utf8),
                  let arr = try JSONSerialization.jsonObject(with: jsonData) as? [[String: Any]],
                  let obj = arr.first else { throw ServiceError.parseError("数据为空") }

            let info = AcademicInfo(
                completedCourses: intValue(obj["courseNum"]) ?? 0,
                failedCourses: intValue(obj["coursePas"]) ?? 0,
                gpa: doubleValue(obj["gpa"]) ?? 0.0,
                averageScore: doubleValue(obj["sspjf_jd"]),
                averageScoreRank: intValue(obj["sspjf_pm"]),
                averageScoreRankPopulation: intValue(obj["sspjf_rs"]),
                pendingCourses: intValue(obj["courseNum_bxqyxd"]) ?? 0,
                currentTerm: stringValue(obj["zxjxjhh"])
            )
            return .success(info)
        } catch {
            logger.error("getAcademicInfo failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription, retryable: true)
        }
    }

    // MARK: - Terms

    func getAllTerms() async -> UniResponse<[TermItem]> {
        if let cached = cachedTerms, cached.success,
           Date().timeIntervalSince(termsCacheTimestamp) < Self.cacheTTL {
            return cached
        }

        do {
            let url = "\(Self.baseURL)/student/courseSelect/calendarSemesterCurriculum/index"
            let client = await connection.client!
            let (data, _) = try await client.get(url)
            let html = String(data: data, encoding: .utf8) ?? ""
            let doc = try SwiftSoup.parse(html)
            guard let select = try doc.select("select#planCode").first() else {
                throw ServiceError.parseError("未找到学期选择框")
            }
            let options = try select.select("option")
            var parsedTerms: [TermItem] = []
            for option in options.array() {
                let code = try option.attr("value")
                guard !code.isEmpty else { continue }
                let rawName = try option.text().trimmingCharacters(in: .whitespaces)
                let name = rawName
                    .replacingOccurrences(of: "春", with: "下")
                    .replacingOccurrences(of: "秋", with: "上")
                let isCurrent = option.hasAttr("selected") || rawName.contains("当前")
                parsedTerms.append(TermItem(termCode: code, termName: name, isCurrent: isCurrent))
            }
            let terms: [TermItem]
            if parsedTerms.contains(where: { $0.isCurrent }) {
                terms = parsedTerms
            } else {
                terms = parsedTerms.enumerated().map { index, term in
                    TermItem(termCode: term.termCode, termName: term.termName, isCurrent: index == 0)
                }
            }
            let result: UniResponse<[TermItem]> = .success(terms)
            cachedTerms = result
            termsCacheTimestamp = Date()
            return result
        } catch {
            logger.error("getAllTerms failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription, retryable: true)
        }
    }

    // MARK: - Scores

    func getTermScore(termCode: String) async -> UniResponse<TermScoreResponse> {
        do {
            let client = await connection.client!
            let preUrl = "\(Self.baseURL)/student/integratedQuery/scoreQuery/allTermScores/index"
            let (preData, _) = try await client.get(preUrl)
            let preHtml = String(data: preData, encoding: .utf8) ?? ""
            guard let pathRange = preHtml.range(of: "/([A-Za-z0-9]+)/allTermScores/data", options: .regularExpression),
                  let match = preHtml[pathRange].range(of: "[A-Za-z0-9]+", options: .regularExpression) else {
                throw ServiceError.parseError("未能提取动态路径")
            }
            let dynamicPath = String(preHtml[pathRange][match])

            let scoreUrl = "\(Self.baseURL)/student/integratedQuery/scoreQuery/\(dynamicPath)/allTermScores/data"
            let (scoreData, _) = try await client.post(scoreUrl, formData: [
                "zxjxjhh": termCode, "kch": "", "kcm": "",
                "pageNum": "1", "pageSize": "100", "sf_request_type": "ajax"
            ], headers: ["Referer": preUrl])

            let body = String(data: scoreData, encoding: .utf8) ?? ""
            guard let jsonData = body.data(using: .utf8),
                  let json = try JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else {
                return .success(TermScoreResponse())
            }
            guard let listData = json["list"] as? [String: Any],
                  let recordsArr = listData["records"] as? [[Any]] else {
                return .success(TermScoreResponse())
            }
            let records = recordsArr.compactMap { parseScoreRecord($0) }
            let total = (listData["pageContext"] as? [String: Any])?["totalCount"] as? Int ?? records.count
            return .success(TermScoreResponse(totalCount: total, records: records))
        } catch {
            logger.error("getTermScore failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription, retryable: true)
        }
    }

    private func parseScoreRecord(_ arr: [Any]) -> ScoreRecord? {
        guard arr.count >= 11 else { return nil }
        func str(_ idx: Int) -> String { "\(arr[idx])" == "<null>" ? "" : "\(arr[idx])" }
        func optStr(_ idx: Int) -> String? { idx < arr.count && "\(arr[idx])" != "<null>" ? "\(arr[idx])" : nil }
        return ScoreRecord(
            sequence: arr[0] as? Int ?? Int(str(0)) ?? 0,
            termId: str(1), courseCode: str(2), courseClass: str(3),
            courseNameCn: str(4), courseNameEn: str(5),
            credits: str(6), hours: arr[7] as? Int ?? Int(str(7)) ?? 0,
            courseType: optStr(8), examType: optStr(9), score: str(10),
            retakeScore: optStr(11), makeupScore: optStr(12)
        )
    }

    // MARK: - Exams

    func getExamInfo() async -> UniResponse<[UnifiedExamInfo]> {
        do {
            let academicResp = await getAcademicInfo()
            guard academicResp.success, let academicData = academicResp.data else {
                throw ServiceError.parseError("无法获取学期信息")
            }
            let termCode = academicData.currentTerm
            let now = Date()
            let cal = Calendar.current
            let startDate = Self.dateString(now)
            let year = cal.component(.year, from: now)
            let endDate = termCode.hasSuffix("1") ? "\(year + 1)-03-30" : "\(year)-09-30"

            let client = await connection.client!
            let preUrl = "\(Self.baseURL)/student/examinationManagement/examPlan/index"
            _ = try await client.get(preUrl)

            let ts = Int(Date().timeIntervalSince1970 * 1000)
            let examUrl = "\(Self.baseURL)/student/examinationManagement/examPlan/detail?start=\(startDate)&end=\(endDate)&_=\(ts)"
            let (examData, _) = try await client.get(examUrl, headers: [
                "Accept": "application/json, text/javascript, */*; q=0.01",
                "X-Requested-With": "XMLHttpRequest"
            ])
            let examBody = String(data: examData, encoding: .utf8) ?? "[]"
            var schoolExams: [UnifiedExamInfo] = []
            if !examBody.trimmingCharacters(in: .whitespaces).isEmpty,
               examBody.trimmingCharacters(in: .whitespaces) != "]",
               let jsonData = examBody.data(using: .utf8),
               let arr = try? JSONSerialization.jsonObject(with: jsonData) as? [[String: Any]] {
                schoolExams = arr.compactMap { parseSchoolExam($0) }
            }

            let otherUrl = "\(Self.baseURL)/student/examinationManagement/othersExamPlan/queryScores?sf_request_type=ajax"
            let (otherData, _) = try await client.post(otherUrl, formData: [
                "zxjxjhh": termCode, "tab": "0", "pageNum": "1", "pageSize": "30"
            ])
            var otherExams: [UnifiedExamInfo] = []
            if let otherBody = String(data: otherData, encoding: .utf8),
               let jsonData = otherBody.data(using: .utf8),
               let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
               let list = json["list"] as? [String: Any],
               let records = list["records"] as? [[Any]] {
                otherExams = records.compactMap { parseOtherExam($0) }
            }

            let all = (schoolExams + otherExams).sorted { ($0.examDate, $0.examTime) < ($1.examDate, $1.examTime) }
            return .success(all)
        } catch {
            logger.error("getExamInfo failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription, retryable: true)
        }
    }

    private func parseSchoolExam(_ obj: [String: Any]) -> UnifiedExamInfo? {
        guard let title = obj["title"] as? String else { return nil }
        let lines = title.split(separator: "\n").map { $0.trimmingCharacters(in: .whitespaces) }
        return UnifiedExamInfo(
            courseName: lines.first ?? "",
            examDate: obj["start"] as? String ?? "",
            examTime: lines.count > 1 ? lines[1] : "",
            examLocation: lines.count > 2 ? lines.dropFirst(2).joined(separator: " ") : "",
            examType: "校统考"
        )
    }

    private func parseOtherExam(_ arr: [Any]) -> UnifiedExamInfo? {
        guard arr.count >= 8 else { return nil }
        func str(_ i: Int) -> String { "\(arr[i])" == "<null>" ? "" : "\(arr[i])" }
        return UnifiedExamInfo(
            courseName: str(2), examDate: str(4), examTime: str(5),
            examLocation: str(6), examType: "其他考试", note: str(7)
        )
    }

    // MARK: - Training Plan

    func getTrainingPlanInfo() async -> UniResponse<TrainingPlanInfo> {
        do {
            let client = await connection.client!
            let url = "\(Self.baseURL)/main/showPyfaInfo?sf_request_type=ajax"
            let (data, _) = try await client.get(url)
            let body = String(data: data, encoding: .utf8) ?? ""
            guard let jsonData = body.data(using: .utf8),
                  let json = try JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
                  let dataList = json["data"] as? [[Any]],
                  let planArr = dataList.first,
                  let planName = planArr.first as? String else {
                throw ServiceError.parseError("无培养方案数据")
            }
            let gradeRegex = try NSRegularExpression(pattern: "(\\d{4})级")
            let grade = gradeRegex.firstMatch(in: planName, range: NSRange(planName.startIndex..., in: planName))
                .flatMap { Range($0.range(at: 1), in: planName) }.map { String(planName[$0]) } ?? ""
            let majorName = planName
                .replacingOccurrences(of: "\\d{4}级", with: "", options: .regularExpression)
                .replacingOccurrences(of: "本科培养方案", with: "")
                .replacingOccurrences(of: "培养方案", with: "")
                .trimmingCharacters(in: .whitespaces)
            return .success(TrainingPlanInfo(planName: planName, majorName: majorName, grade: grade))
        } catch {
            logger.error("getTrainingPlanInfo failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription, retryable: true)
        }
    }

    private static func dateString(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }
}

private func stringValue(_ value: Any?, fallback: String = "") -> String {
    guard let value, !(value is NSNull) else { return fallback }
    return String(describing: value)
}

private func intValue(_ value: Any?) -> Int? {
    guard let value, !(value is NSNull) else { return nil }
    if let int = value as? Int { return int }
    if let double = value as? Double { return Int(double) }
    if let string = value as? String { return Int(string) }
    return nil
}

private func doubleValue(_ value: Any?) -> Double? {
    guard let value, !(value is NSNull) else { return nil }
    if let double = value as? Double { return double }
    if let int = value as? Int { return Double(int) }
    if let string = value as? String { return Double(string) }
    return nil
}

enum ServiceError: Error, LocalizedError {
    case emptyResponse
    case parseError(String)
    case sessionExpired

    var errorDescription: String? {
        switch self {
        case .emptyResponse: return "响应为空"
        case .parseError(let msg): return msg
        case .sessionExpired: return "会话已过期"
        }
    }
}

File: android/app/src/main/java/tech/loveace/appv3/data/service/JWCService.kt
package tech.loveace.appv3.data.service

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.*
import org.jsoup.Jsoup
import tech.loveace.appv3.data.model.*
import tech.loveace.appv3.data.network.AUFEConnection

/**
 * 教务系统服务 - 学业信息、成绩、考试、学期、培养方案
 */
class JWCService(private val connection: AUFEConnection) {

    private val termsMutex = Mutex()
    private var cachedTerms: UniResponse<List<TermItem>>? = null
    private var termsCacheTimestamp: Long = 0

    // ==================== Academic Info ====================
    suspend fun getAcademicInfo(): UniResponse<AcademicInfo> = withContext(Dispatchers.IO) {
        try {
            val url = "$BASE_URL/main/academicInfo?sf_request_type=ajax"
            val response = connection.client.post(
                url,
                formData = mapOf("flag" to ""),
                headers = mapOf(
                    "Accept" to "application/json, text/javascript, */*; q=0.01",
                    "X-Requested-With" to "XMLHttpRequest",
                ),
            )
            val body = response.body?.string() ?: throw Exception("响应为空")
            val json = Json.parseToJsonElement(body)
            val arr = json.jsonArray
            if (arr.isEmpty()) throw Exception("数据为空")
            val obj = arr[0].jsonObject
            val info = AcademicInfo(
                completedCourses = obj["courseNum"]?.jsonPrimitive?.intOrNull ?: 0,
                failedCourses = obj["coursePas"]?.jsonPrimitive?.intOrNull ?: 0,
                gpa = obj["gpa"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
                averageScore = obj["sspjf_jd"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
                averageRank = obj["sspjf_pm"]?.jsonPrimitive?.intOrNull ?: 0,
                averageRankTotal = obj["sspjf_rs"]?.jsonPrimitive?.intOrNull ?: 0,
                pendingCourses = obj["courseNum_bxqyxd"]?.jsonPrimitive?.intOrNull ?: 0,
                currentTerm = obj["zxjxjhh"]?.jsonPrimitive?.contentOrNull ?: "",
            )
            UniResponse.success(info)
        } catch (e: Exception) {
            Log.e(TAG, "getAcademicInfo failed", e)
            UniResponse.failure(e.message ?: "获取学业信息失败", retryable = true)
        }
    }

    // ==================== Terms ====================
    suspend fun getAllTerms(): UniResponse<List<TermItem>> = termsMutex.withLock {
        val now = System.currentTimeMillis()
        val cached = cachedTerms
        if (cached?.success == true && now - termsCacheTimestamp < CACHE_TTL_MS) {
            Log.d(TAG, "Cache hit for terms")
            return@withLock cached
        }

        val result = withContext(Dispatchers.IO) {
            try {
                val url = "$BASE_URL/student/courseSelect/calendarSemesterCurriculum/index"
                val response = connection.client.get(url)
                val html = response.body?.string() ?: throw Exception("响应为空")
                val doc = Jsoup.parse(html)
                val select = doc.selectFirst("select#planCode") ?: throw Exception("未找到学期选择框")
                val options = select.select("option")
                val parsedTerms = options.mapNotNull { option ->
                    val code = option.attr("value").takeIf { it.isNotEmpty() } ?: return@mapNotNull null
                    val rawName = option.text().trim()
                    val name = rawName.replace("春", "下").replace("秋", "上")
                    val isCurrent = option.hasAttr("selected") || rawName.contains("当前")
                    TermItem(termCode = code, termName = name, isCurrent = isCurrent)
                }
                val terms = if (parsedTerms.any { it.isCurrent }) {
                    parsedTerms
                } else {
                    parsedTerms.mapIndexed { index, term ->
                        if (index == 0) term.copy(isCurrent = true) else term
                    }
                }
                UniResponse.success(terms)
            } catch (e: Exception) {
                Log.e(TAG, "getAllTerms failed", e)
                UniResponse.failure(e.message ?: "获取学期列表失败", retryable = true)
            }
        }

        if (result.success) {
            cachedTerms = result
            termsCacheTimestamp = System.currentTimeMillis()
        }

        result
    }

    // ==================== Scores ====================
    suspend fun getTermScore(termCode: String): UniResponse<TermScoreResponse> = withContext(Dispatchers.IO) {
        try {
            // Step 1: Get dynamic path
            val preUrl = "$BASE_URL/student/integratedQuery/scoreQuery/allTermScores/index"
            val preResponse = connection.client.get(preUrl)
            val preHtml = preResponse.body?.string() ?: throw Exception("页面为空")
            val pathMatch = Regex("/([A-Za-z0-9]+)/allTermScores/data").find(preHtml)
                ?: throw Exception("未能提取动态路径")
            val dynamicPath = pathMatch.groupValues[1]

            // Step 2: Fetch scores
            val scoreUrl = "$BASE_URL/student/integratedQuery/scoreQuery/$dynamicPath/allTermScores/data"
            val scoreResponse = connection.client.post(
                scoreUrl,
                formData = mapOf(
                    "zxjxjhh" to termCode, "kch" to "", "kcm" to "",
                    "pageNum" to "1", "pageSize" to "100", "sf_request_type" to "ajax",
                ),
                headers = mapOf("Referer" to preUrl),
            )
            val body = scoreResponse.body?.string() ?: throw Exception("成绩数据为空")
            val json = Json.parseToJsonElement(body).jsonObject
            val listData = json["list"]?.jsonObject ?: return@withContext UniResponse.success(
                TermScoreResponse(0, emptyList())
            )
            val recordsArr = listData["records"]?.jsonArray ?: return@withContext UniResponse.success(
                TermScoreResponse(0, emptyList())
            )
            val records = recordsArr.mapNotNull { parseScoreRecord(it.jsonArray) }
            val total = listData["pageContext"]?.jsonObject?.get("totalCount")?.jsonPrimitive?.intOrNull ?: records.size
            UniResponse.success(TermScoreResponse(total, records))
        } catch (e: Exception) {
            Log.e(TAG, "getTermScore failed", e)
            UniResponse.failure(e.message ?: "获取成绩失败", retryable = true)
        }
    }

    private fun parseScoreRecord(arr: JsonArray): ScoreRecord? {
        if (arr.size < 11) return null
        return ScoreRecord(
            sequence = arr[0].jsonPrimitive.intOrNull ?: 0,
            termId = arr[1].jsonPrimitive.contentOrNull ?: "",
            courseCode = arr[2].jsonPrimitive.contentOrNull ?: "",
            courseClass = arr[3].jsonPrimitive.contentOrNull ?: "",
            courseNameCn = arr[4].jsonPrimitive.contentOrNull ?: "",
            courseNameEn = arr[5].jsonPrimitive.contentOrNull ?: "",
            credits = arr[6].jsonPrimitive.contentOrNull ?: "0",
            hours = arr[7].jsonPrimitive.contentOrNull?.toIntOrNull() ?: 0,
            courseType = arr.getOrNull(8)?.jsonPrimitive?.contentOrNull,
            examType = arr.getOrNull(9)?.jsonPrimitive?.contentOrNull,
            score = arr[10].jsonPrimitive.contentOrNull ?: "",
            retakeScore = arr.getOrNull(11)?.jsonPrimitive?.contentOrNull,
            makeupScore = arr.getOrNull(12)?.jsonPrimitive?.contentOrNull,
        )
    }

    // ==================== Exams ====================
    suspend fun getExamInfo(): UniResponse<List<UnifiedExamInfo>> = withContext(Dispatchers.IO) {
        try {
            val academicResp = getAcademicInfo()
            if (!academicResp.success || academicResp.data == null) throw Exception("无法获取学期信息")
            val termCode = academicResp.data.currentTerm
            val now = java.time.LocalDate.now()
            val startDate = now.toString()
            val endDate = if (termCode.endsWith("1")) "${now.year + 1}-03-30" else "${now.year}-09-30"

            // Fetch school exams
            val preUrl = "$BASE_URL/student/examinationManagement/examPlan/index"
            val preResp = connection.client.get(preUrl)
            val preHtml = preResp.body?.string() ?: ""
            val seatInfos = parseExamSeatInfo(preHtml)
            val ts = System.currentTimeMillis()
            val examUrl = "$BASE_URL/student/examinationManagement/examPlan/detail?start=$startDate&end=$endDate&_=$ts"
            val examResp = connection.client.get(examUrl, headers = mapOf(
                "Accept" to "application/json, text/javascript, */*; q=0.01",
                "X-Requested-With" to "XMLHttpRequest",
            ))
            val examBody = examResp.body?.string() ?: "[]"
            val schoolExams = if (examBody.trim() == "]" || examBody.isBlank()) emptyList()
            else try {
                Json.parseToJsonElement(examBody).jsonArray.mapNotNull { parseSchoolExam(it.jsonObject, seatInfos) }
            } catch (_: Exception) { emptyList() }

            // Fetch other exams
            val otherUrl = "$BASE_URL/student/examinationManagement/othersExamPlan/queryScores?sf_request_type=ajax"
            val otherResp = connection.client.post(otherUrl, formData = mapOf(
                "zxjxjhh" to termCode, "tab" to "0", "pageNum" to "1", "pageSize" to "30",
            ))
            val otherBody = otherResp.body?.string() ?: "{}"
            val otherExams = try {
                val json = Json.parseToJsonElement(otherBody).jsonObject
                val records = json["records"]?.jsonArray
                    ?: json["list"]?.jsonObject?.get("records")?.jsonArray
                    ?: JsonArray(emptyList())
                records.mapNotNull { parseOtherExam(it) }
            } catch (_: Exception) { emptyList() }

            val all = (schoolExams + otherExams).sortedWith(compareBy({ it.examDate }, { it.examTime }))
            UniResponse.success(all)
        } catch (e: Exception) {
            Log.e(TAG, "getExamInfo failed", e)
            UniResponse.failure(e.message ?: "获取考试信息失败", retryable = true)
        }
    }

    private data class ExamSeatInfo(val courseName: String, val seatNumber: String)

    private fun parseExamSeatInfo(html: String): List<ExamSeatInfo> {
        if (html.isBlank()) return emptyList()
        val doc = Jsoup.parse(html)
        return doc.select("div.widget-box").mapNotNull { box ->
            val titleElement = box.selectFirst("h5.widget-title") ?: return@mapNotNull null
            val mainElement = box.selectFirst("div.widget-main") ?: return@mapNotNull null
            val courseTitle = titleElement.text().trim()
            val courseName = Regex("[）)](.+)$").find(courseTitle)?.groupValues?.get(1)?.trim() ?: courseTitle
            val seatNumber = Regex("座位号[：:](.+?)(?:准考证号|$)")
                .find(mainElement.text().trim())
                ?.groupValues
                ?.get(1)
                ?.trim()
                ?: return@mapNotNull null

            if (seatNumber.isEmpty()) null else ExamSeatInfo(courseName, seatNumber)
        }
    }

    private fun parseSchoolExam(obj: JsonObject, seatInfos: List<ExamSeatInfo>): UnifiedExamInfo? {
        val title = obj["title"]?.jsonPrimitive?.contentOrNull ?: return null
        val lines = title.split("\n").map { it.trim() }
        val courseName = lines.getOrElse(0) { "" }
        val seatNumber = seatInfos.firstOrNull { it.courseName == courseName }?.seatNumber.orEmpty()
        return UnifiedExamInfo(
            courseName = courseName,
            examDate = obj["start"]?.jsonPrimitive?.contentOrNull ?: "",
            examTime = lines.getOrElse(1) { "" },
            examLocation = lines.drop(2).joinToString(" "),
            examType = "校统考",
            note = if (seatNumber.isEmpty()) "" else "座位号: $seatNumber",
        )
    }

    private fun parseOtherExam(element: JsonElement): UnifiedExamInfo? = when (element) {
        is JsonArray -> parseOtherExamArray(element)
        is JsonObject -> parseOtherExamObject(element)
        else -> null
    }

    private fun parseOtherExamArray(arr: JsonArray): UnifiedExamInfo? {
        if (arr.size < 8) return null
        return UnifiedExamInfo(
            courseName = arr[2].jsonPrimitive.contentOrNull ?: "",
            examDate = arr[4].jsonPrimitive.contentOrNull ?: "",
            examTime = arr[5].jsonPrimitive.contentOrNull ?: "",
            examLocation = arr[6].jsonPrimitive.contentOrNull ?: "",
            examType = "其他考试",
            note = arr.getOrNull(7)?.jsonPrimitive?.contentOrNull ?: "",
        )
    }

    private fun parseOtherExamObject(obj: JsonObject): UnifiedExamInfo? {
        val courseName = obj.string("KCM")
        val examDate = obj.string("KSRQ")
        val examTime = obj.string("KSSJ")
        if (courseName.isEmpty() && examDate.isEmpty() && examTime.isEmpty()) return null
        return UnifiedExamInfo(
            courseName = courseName,
            examDate = examDate,
            examTime = examTime,
            examLocation = obj.string("KSDD"),
            examType = "其他考试",
            note = obj.string("BZ"),
        )
    }

    // ==================== Training Plan ====================
    suspend fun getTrainingPlanInfo(): UniResponse<TrainingPlanInfo> = withContext(Dispatchers.IO) {
        try {
            val url = "$BASE_URL/main/showPyfaInfo?sf_request_type=ajax"
            val response = connection.client.get(url)
            val body = response.body?.string() ?: throw Exception("响应为空")
            val json = Json.parseToJsonElement(body).jsonObject
            val dataList = json["data"]?.jsonArray ?: throw Exception("无培养方案数据")
            val planArr = dataList[0].jsonArray
            val planName = planArr[0].jsonPrimitive.content
            val gradeMatch = Regex("(\\d{4})级").find(planName)
            val grade = gradeMatch?.groupValues?.get(1) ?: ""
            val majorName = planName.replace(Regex("\\d{4}级"), "")
                .replace("本科培养方案", "").replace("培养方案", "").trim()
            UniResponse.success(TrainingPlanInfo(planName, majorName, grade))
        } catch (e: Exception) {
            Log.e(TAG, "getTrainingPlanInfo failed", e)
            UniResponse.failure(e.message ?: "获取培养方案失败", retryable = true)
        }
    }

    companion object {
        private const val TAG = "JWCService"
        private const val CACHE_TTL_MS = 30_000L
        const val BASE_URL = "http://jwcxk2-aufe-edu-cn.vpn2.aufe.edu.cn:8118"
    }
}

private fun JsonObject.string(key: String): String = this[key]?.jsonPrimitive?.contentOrNull ?: ""
